### PR TITLE
[NNC] Bounds Inference: make inferred bounds respect gaps

### PR DESCRIPTION
**Pull request by nickgg**
A heavy refactor of bounds inference to fix some issues and bugs blocking using it to analyze cross thread interactions:
* We were merging all accesses to a Buf into a single bounds info entry, even if they did not overlap. E.g. if we accessed a[0:2] and a[5:6] we would merge that into a bound of a[0:6]. I've changed this behaviour to merge only overlapping bounds.
* We were not separating bounds of different kinds (e.g. Load vs Store) and would merge a Store bounds into a Load bounds, losing the information about what kind of access it was. E.g. this loop would produce bounds: [{Load, 0, 10}] and now produces bounds [{Load, 0, 9}, {Store, 1, 10}]:
```
for i in 1 to 10...
  x[i] = x[i-1]
```
* Both ComputeAt and Rfactor relied on the overzealous merging and only used a single entry in the bounds list to determine the bounds of temporary buffers they created, which could result in temporary buffers allocated smaller than accesses to them. I've fixed Rfactor, but *not* ComputeAt - however all ComputeAt tests still pass (may require loop fusion to trigger this issue) - I will come back to it.

Being more precise about bounds is more complex, rather than taking the minimum of starts and maximum of stops we now need to determine if two bounds overlap or are adjacent. There are many edge cases and so I've added a bunch of test coverage of the merging method.

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/bounds_inference.cpp [hunk 1]
```diff
@@ -128,36 +192,49 @@ static const Expr* simplifyMax(const Expr* a, const Expr* b) {
...
+  for (auto& pair : unmerged) {
+    std::vector<TensorAccessBoundsInfo>& res_vec = res[pair.first];
+    for (const auto& v : pair.second) {
+      bool found = false;
+      for (auto& v2 : res_vec) {
+        if (v2.kind != v.kind) {
+          continue;
+        }
+
+        assert(v.start.size() == v.stop.size());
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/bounds_inference.cpp:206]: `TORCH_INTERNAL_ASSERT` should probably be used instead of `assert`.

File: torch/csrc/jit/tensorexpr/bounds_inference.cpp [hunk 2]
```diff
@@ -128,36 +192,49 @@ static const Expr* simplifyMax(const Expr* a, const Expr* b) {
...
+        assert(v2.start.size() == v2.stop.size());
+        assert(v.start.size() == v2.start.size());
+
+        std::vector<const Expr*> start;
+        std::vector<const Expr*> stop;
+        bool fail = false;
+        for (unsigned int i = 0; i < v.start.size(); ++i) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/bounds_inference.cpp:213]: The code is a bit hard to read, maybe use more descriptive names for variables and/or specify types instead of using auto?

File: torch/csrc/jit/tensorexpr/bounds_inference.cpp [hunk 3]
```diff
@@ -128,36 +192,49 @@ static const Expr* simplifyMax(const Expr* a, const Expr* b) {
...
-    for (size_t i = 0; i < old_t.start.size(); i++) {
-      new_t.start[i] = simplifyMin(old_t.start[i], new_t.start[i]);
-    }
-    for (size_t i = 0; i < old_t.stop.size(); i++) {
-      new_t.stop[i] = simplifyMax(old_t.stop[i], new_t.stop[i]);
+  for (auto& pair : unmerged) {
+    std::vector<TensorAccessBoundsInfo>& res_vec = res[pair.first];
+    for (const auto& v : pair.second) {
+      bool found = false;
+      for (auto& v2 : res_vec) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/bounds_inference.cpp:201]: Do I understand it correctly that now the algorithm has become quadratic? Not that I expect it to be a problem, but want to double check whether I read the code correctly.

File: torch/csrc/jit/tensorexpr/loopnest.cpp [hunk 4]
```diff
@@ -1703,24 +1705,41 @@ void LoopNest::rfactor(
...
+        "Hit undefined behavior in rfactor -- couldn't infer bounds.");
+  }
+
+  std::vector<const Expr*> starts;
+  std::vector<const Expr*> stops;
+
+  for (const TensorAccessBoundsInfo& p : bounds_it->second) {
+    for (size_t i = 0; i < p.start.size(); i++) {
+      if (starts.size() <= i) {
+        starts.push_back(p.start[i]);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/loopnest.cpp:1720]: Nit: I think a comment could be helpful before the loop.

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 5]
```diff
@@ -942,6 +942,50 @@ const Expr* PolynomialTransformer::mutate(const Div* v) {
...
+  }
+
+  if (immediateAs<int>(diff) > 0) {
+    return lhs_new;
+  }
+
+  return rhs_new;
+}
+
+const Expr* PolynomialTransformer::mutate(const Min* v) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_simplifier.cpp:967]: Should we have dedicated tests for these simplifications?

File: torch/csrc/jit/tensorexpr/bounds_inference.cpp [hunk 6]
```diff
@@ -27,36 +27,39 @@ class BoundsInference : public IRVisitor {
...
-      // all cases.
-      const Expr* old_start = access.start[j];
-      const Expr* old_stop = access.stop[j];
-      const Expr* new_start = Substitute(old_start, {{v->var(), v->start()}});
-      const Expr* new_stop =
-          Substitute(old_stop, {{v->var(), new Sub(v->stop(), new IntImm(1))}});
-      access.start[j] = IRSimplifier::simplify(new_start);
-      access.stop[j] = IRSimplifier::simplify(new_stop);
+  for (auto& pair : accesses_) {
+    for (TensorAccessBoundsInfo& access : pair.second) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/bounds_inference.cpp:46]: The main reason the original code used a vector instead of a map is to prevent non-determinism (the order in vector was defined by the order of appearance in the statement). I don't think we're using that order anywhere right now, but if we replace vector with a map we should be on a watch out for changes that could accidentally iterate over the map and depend on the order (e.g. if we create some temporary variables as we see entries in the map, it would be quite annoying to see difference in debug dumps from run to run).

**Review by ZolotukhinM — APPROVED** [review]
Overall looks good, I left some comments inline.

File: test/cpp/tensorexpr/test_loopnest.cpp [hunk 7]
```diff
@@ -645,295 +645,6 @@ void testScheduleDynamicShape2D() {
...
-    }
-    if (ref[i].second >= 0) {
-      auto stop_imm = dynamic_cast<const IntImm*>(access_info.stop[i]);
-      ASSERT_TRUE(stop_imm);
-      ASSERT_EQ(stop_imm->value(), ref[i].second);
-    }
-  }
-}
-
-void testBoundsInference_1() {
```
[the remark this answers is not in the document]
Comment (nickgg) [reply test/cpp/tensorexpr/test_loopnest.cpp:677]: Not verbatim, they were all updated for the inferBounds interface change, but yeah they were all moved.

File: torch/csrc/jit/tensorexpr/bounds_inference.cpp (hunk 6 above)
Comment (nickgg) [reply torch/csrc/jit/tensorexpr/bounds_inference.cpp:46]: The map is keyed with the Buf, the accesses themselves are still ordered in the vector (although merging always changes that). Do you really see a situation where where we do rely on the order buffers appear in this list?

File: torch/csrc/jit/tensorexpr/bounds_inference.cpp (hunk 3 above)
Comment (nickgg) [reply torch/csrc/jit/tensorexpr/bounds_inference.cpp:201]: Yeah it's n^2, specifically (n^2-n)/2 in worst case. I think n, which is the number of unmerged accesses to a specific buffer, will be small enough for this to be ok.

File: torch/csrc/jit/tensorexpr/bounds_inference.cpp (hunk 2 above)
Comment (nickgg) [reply torch/csrc/jit/tensorexpr/bounds_inference.cpp:213]: fair, I'll clean it up a bit.

File: torch/csrc/jit/tensorexpr/bounds_inference.h [hunk 8]
```diff
@@ -16,18 +17,19 @@ class Stmt;
...
 
 struct TORCH_API TensorAccessBoundsInfo {
-  const Buf* buf;
   TensorAccessKind kind;
   std::vector<const Expr*> start;
   std::vector<const Expr*> stop;
 };
 
-using BoundsInfo = std::vector<TensorAccessBoundsInfo>;
+using BoundsInfo = std::map<const Buf*, std::vector<TensorAccessBoundsInfo>>;
```
[the remark this answers is not in the document]
Comment (nickgg) [reply torch/csrc/jit/tensorexpr/bounds_inference.h:25]: map is a little more space efficient for small numbers of elements, but it's pretty marginal.
